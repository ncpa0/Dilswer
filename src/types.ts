import type { BasicDataTypes, dataTypeSymbol } from "./type-def-helpers";
import type { ValueOf } from "./type-utils";

export type ArrayOf<DT extends AllDataTypes[] = any[]> = {
  readonly [dataTypeSymbol]: true;
  readonly arrayOf: DT;
};

export type RecordOf<TS extends RecordTypeSchema = RecordTypeSchema> = {
  readonly [dataTypeSymbol]: true;
  readonly recordOf: TS;
};

export type SetOf<DT extends AllDataTypes[] = any[]> = {
  readonly [dataTypeSymbol]: true;
  readonly setOf: DT;
};

export type OneOf<DT extends AllDataTypes[] = any[]> = {
  readonly [dataTypeSymbol]: true;
  readonly oneOf: DT;
};

export type Literal<
  V extends string | number | boolean = string | number | boolean
> = {
  readonly [dataTypeSymbol]: true;
  readonly literal: V;
};

export type Enum<E = any> = {
  readonly [dataTypeSymbol]: true;
  readonly enumInstance: E;
};

export type EnumMember<M = any> = {
  readonly [dataTypeSymbol]: true;
  readonly enumMember: M;
};

export type BasicDataType = ValueOf<typeof BasicDataTypes>;

export type ComplexDataType =
  | ArrayOf
  | RecordOf
  | SetOf
  | OneOf
  | Literal
  | Enum
  | EnumMember;

export type AllDataTypes = BasicDataType | ComplexDataType;

export type FieldDescriptor = {
  readonly required?: boolean;
  readonly type: AllDataTypes;
};

export interface RecordTypeSchema {
  readonly [key: string]: FieldDescriptor | AllDataTypes;
}
