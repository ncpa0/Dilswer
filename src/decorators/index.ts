export { ClassWithValidation, validate } from "./validation-decorator";
