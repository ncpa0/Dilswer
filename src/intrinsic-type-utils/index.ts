export * from "@Intrinsic/and";
export * from "@Intrinsic/exclude";
export * from "@Intrinsic/omit";
export * from "@Intrinsic/partial";
export * from "@Intrinsic/pick";
export * from "@Intrinsic/required";
