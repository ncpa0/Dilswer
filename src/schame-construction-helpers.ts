import { AllDataTypes } from ".";
import type {
  ArrayOf,
  Enum,
  EnumMember,
  Literal,
  OneOf,
  RecordOf,
  SetOf,
  TypeSchema,
} from "./types";

export const BasicDataTypes = {
  Unknown: "unknown",
  String: "string",
  Number: "number",
  Boolean: "boolean",
  Symbol: "symbol",
  Function: "function",
  Null: "null",
  Undefined: "undefined",
} as const;

export const DataType = {
  ...BasicDataTypes,
  RecordOf<TS extends TypeSchema>(args: TS): RecordOf<TS> {
    return { recordOf: args };
  },
  ArrayOf<DT extends AllDataTypes[]>(...args: DT): ArrayOf<DT> {
    return { arrayOf: args };
  },
  SetOf<DT extends AllDataTypes[]>(...args: DT): SetOf<DT> {
    return { setOf: args };
  },
  OneOf<DT extends AllDataTypes[]>(...args: DT): OneOf<DT> {
    return { oneOf: args };
  },
  Literal<V extends string | number | boolean>(value: V): Literal<V> {
    return { literal: value };
  },
  EnumMember<M extends number | string>(enumMember: M): EnumMember<M> {
    return { enumMember };
  },
  Enum<T extends string, TEnumValue extends string | number>(enumInstance: {
    [key in T]: TEnumValue;
  }): Enum<TEnumValue> {
    // @ts-expect-error
    return { enumInstance };
  },
};
