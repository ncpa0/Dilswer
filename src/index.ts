import {
  createChecker,
  createValidator,
} from "./checker-machine/create-checker";
import {
  createTypeGuardedFunction,
  createValidatedFunction,
} from "./checker-machine/create-validated-function";
import { ensureDataType } from "./checker-machine/ensure-data-type";
import { DataType } from "./schame-construction-helpers";
import { And, Exclude, Omit, Partial, Pick, Required } from "./utilities";

export default {
  createChecker,
  createValidator,
  createTypeGuardedFunction,
  createValidatedFunction,
  ensureDataType,
  And,
  Exclude,
  Omit,
  Partial,
  Pick,
  Required,
  DataType,
};

export {
  createChecker,
  createValidator,
} from "./checker-machine/create-checker";
export {
  createTypeGuardedFunction,
  createValidatedFunction,
} from "./checker-machine/create-validated-function";
export { ensureDataType } from "./checker-machine/ensure-data-type";
export { DataType } from "./schame-construction-helpers";
export { GetDataType } from "./type-utils";
export {
  AllDataTypes,
  ArrayOf,
  BasicDataType,
  ComplexDataType,
  Enum,
  EnumMember,
  FieldDescriptor,
  Literal,
  OneOf,
  RecordOf,
  RecordTypeSchema as TypeSchema,
  SetOf,
} from "./types";
export * from "./utilities";
