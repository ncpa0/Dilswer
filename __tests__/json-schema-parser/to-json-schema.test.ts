import Ajv from "ajv";
import { OptionalField, toJsonSchema, Type } from "../../src";

enum Enum {
  A = "A",
  B = "B",
  C = "C",
}

const testDt = Type.Record({
  uuid: Type.String,
  number: Type.Number,
  bool: Type.Boolean,
  stringSet: Type.Set(Type.String),
  stringArray: Type.Array(Type.String),
  unionArray: Type.Array(
    Type.Record({ foo: Type.String }),
    Type.Record({ bar: Type.String }),
  ),
  tuple: Type.Tuple(Type.String, Type.Number),
  literalString: Type.Literal("literal"),
  literalNumber: Type.Literal(1),
  literalBoolean: Type.Literal(true),
  enum: Type.Enum(Enum),
  bMember: Type.EnumMember(Enum.B),
  recordIntersection: Type.AllOf(
    Type.Record({
      a: Type.String,
    }),
    Type.Record({
      b: Type.Number,
    }),
  ),
  dict: Type.Dict(Type.String),
  oneof: Type.OneOf(
    Type.Record({ foo: Type.String }),
    Type.Record({ bar: Type.String }),
  ),
  symbol: Type.Symbol,
  undef: Type.Undefined,
  customValidator: Type.Custom((value: any): value is any => true),
  stringMatching: Type.StringMatching(/^foo$/),
  optionalSelfCopy: {
    required: false,
    type: Type.Record({
      id: { required: false, type: Type.String },
      number: { required: false, type: Type.Number },
      bool: { required: false, type: Type.Boolean },
      stringSet: { required: false, type: Type.Set(Type.String) },
      stringArray: { required: false, type: Type.Array(Type.String) },
      unionArray: {
        required: false,
        type: Type.Array(
          Type.Record({ foo: Type.String }),
          Type.Record({ bar: Type.String }),
        ),
      },
      tuple: {
        required: false,
        type: Type.Tuple(
          Type.Record({
            id: { required: false, type: Type.String },
          }),
          Type.Literal("separator"),
          Type.Boolean,
        ),
      },
      literalString: { required: false, type: Type.Literal("literal") },
      literalNumber: { required: false, type: Type.Literal(1) },
      literalBoolean: { required: false, type: Type.Literal(true) },
      enum: { required: false, type: Type.Enum(Enum) },
      bMember: { required: false, type: Type.EnumMember(Enum.B) },
      recordIntersection: {
        required: false,
        type: Type.AllOf(
          Type.Record({
            a: Type.String,
          }),
          Type.Record({
            b: Type.Number,
          }),
        ),
      },
      dict: { required: false, type: Type.Dict(Type.String) },
      oneof: {
        required: false,
        type: Type.OneOf(
          Type.Record({ foo: Type.String }),
          Type.Record({ bar: Type.String }),
        ),
      },
      stringMatching: {
        required: false,
        type: Type.StringMatching(/foo.+[0-9]$/gi),
      },
    }),
  },
});

describe("toJsonSchema", () => {
  it("should parse to a valid JSONSchema when set to 'omit'", () => {
    const schema = toJsonSchema(testDt, { incompatibleTypes: "omit" }, false);

    expect(schema).toMatchSnapshot();
    expect(schema).toBeDefined();

    const isValid = new Ajv().validateSchema(schema!);

    expect(isValid).toEqual(true);
  });

  it("should parse to a valid JSONSchema when set to 'set-as-any'", () => {
    const schema = toJsonSchema(
      testDt,
      { incompatibleTypes: "set-as-any" },
      false,
    );

    expect(schema).toBeDefined();
    expect(schema).toMatchSnapshot();

    const isValid = new Ajv().validateSchema(schema!);

    expect(isValid).toEqual(true);
  });

  it("should throw an error when attempting to parse incompatible types when set to 'throw'", () => {
    expect(() => toJsonSchema(testDt, { incompatibleTypes: "throw" }))
      .toThrow();
  });

  it("should set the specified additionalProperty", () => {
    const schema = toJsonSchema(
      testDt,
      { incompatibleTypes: "omit", additionalProperties: false },
      false,
    );

    expect(schema).toMatchSnapshot();
    expect(schema).toBeDefined();

    const isValid = new Ajv().validateSchema(schema!);

    expect(isValid).toEqual(true);
  });

  it("should use the custom parsers", () => {
    const dt = Type.Record({
      custom: Type.Custom((value: any): value is any => true),
      function: Type.Function,
      set: Type.Set(Type.String),
      symbol: Type.Symbol,
      undefined: Type.Undefined,
    });

    const schema = toJsonSchema(
      dt,
      {
        customParser: {
          Custom(validateFunction, original, options) {
            return {
              title: "Custom parser schema for validateFunction",
            };
          },
          Function(dataType, options) {
            return {
              title: "Custom parser schema for Function DataType",
            };
          },
          Set(setItemsSchemas, original, options) {
            return {
              title: "Custom parser schema for Set DataType",
            };
          },
          Symbol(dataType, options) {
            return {
              title: "Custom parser schema for Symbol DataType",
            };
          },
          Undefined(dataType, options) {
            return {
              title: "Custom parser schema for Undefined DataType",
            };
          },
        },
      },
      false,
    );

    expect(schema).toMatchSnapshot();
    expect(schema).toBeDefined();

    const isValid = new Ajv().validateSchema(schema!);

    expect(isValid).toEqual(true);
  });

  it("should copy the properties from metadata", () => {
    const dt = Type.Record({
      date: Type.String.setFormat("date-time"),
      timestamp: Type.Int.setDescription("UNIX timestamp.").setTitle(
        "Creation Timestamp",
      ),
      list: Type.Array(Type.String),
      dictionary: OptionalField(
        Type.Record({ foo: Type.Unknown }).setDescription(
          "This field is optional.",
        ),
      ),
    }).setTitle("Top Record");

    dt.recordOf.list
      .setTitle("List of strings")
      .setDescription("This is a list of strings");

    const schema = toJsonSchema(dt, { incompatibleTypes: "set-as-any" });

    expect(schema).toMatchSnapshot();
  });

  describe("should work correctly with circular type", () => {
    it("simple circular record", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Array(self),
        })
      );

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("simple circular named record", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Array(self),
        }).setTitle("Node")
      );

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("immediately self-referencing record", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          self: OptionalField(self),
        }).setTitle("SelfReferencingRecord")
      );

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("simple circular array", () => {
      const typeDef = Type.Recursive((self) => Type.Array(Type.String, self));

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("deeply nested structure", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Record({
            type: Type.String,
            container: Type.Record({
              name: Type.String,
              children: Type.Array(self),
            }),
          }),
        }).setTitle("Node")
      );

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("deeply nested structure of named types", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Array(
            Type.Record({
              name: Type.Literal("span"),
              children: Type.Array(self),
            }).setTitle("SpanNode"),
          ).setTitle("SpanNodeList"),
        }).setTitle("Node")
      );

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("deeply nested circular types", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Recursive((self2) =>
            Type.Array(
              self,
              Type.Record({
                nested: Type.Literal(true),
                items: self2,
              }),
            )
          ),
        })
      );

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("works with dicts", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: Type.Dict(self),
        })
      );

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("works with direct dicts", () => {
      const typeDef = Type.Recursive((self) => Type.Dict(self, Type.String));

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("works with tuples", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: OptionalField(
            Type.OneOf(
              Type.Tuple(self),
              Type.Tuple(self, self),
              Type.Tuple(self, self, self),
            ),
          ),
        })
      );

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("works with direct tuples", () => {
      const typeDef = Type.Recursive((self) => Type.Tuple(self, Type.String));

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("works with unions", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: OptionalField(
            Type.OneOf(
              self,
              Type.Record({
                name: Type.String,
                children: self,
              }),
            ),
          ),
        })
      );

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("works with direct unions", () => {
      const typeDef = Type.Recursive((self) => Type.OneOf(self, Type.String));

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });

    it("works with intersections", () => {
      const typeDef = Type.Recursive((self) =>
        Type.Record({
          name: Type.String,
          children: OptionalField(
            Type.AllOf(
              self,
              Type.Record({
                type: Type.String,
              }),
            ),
          ),
        })
      );

      const schema = toJsonSchema(typeDef, {}, false);

      expect(schema).toMatchSnapshot();

      const isValid = new Ajv().validateSchema(schema!);

      expect(isValid).toEqual(true);
    });
  });
});
