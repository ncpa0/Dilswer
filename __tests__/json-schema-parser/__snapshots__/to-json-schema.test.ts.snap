// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toJsonSchema should copy the properties from metadata 1`] = `
{
  "$schema": "http://json-schema.org/draft-06/schema#",
  "properties": {
    "date": {
      "format": "date-time",
      "type": "string",
    },
    "dictionary": {
      "description": "This field is optional.",
      "properties": {
        "foo": {},
      },
      "required": [
        "foo",
      ],
      "type": "object",
    },
    "list": {
      "description": "This is a list of strings",
      "items": [
        {
          "type": "string",
        },
      ],
      "title": "List of strings",
      "type": "array",
    },
    "timestamp": {
      "description": "UNIX timestamp.",
      "title": "Creation Timestamp",
      "type": "integer",
    },
  },
  "required": [
    "date",
    "timestamp",
    "list",
  ],
  "title": "Top Record",
  "type": "object",
}
`;

exports[`toJsonSchema should parse to a valid JSONSchema when set to 'omit' 1`] = `
{
  "properties": {
    "bMember": {
      "enum": [
        "B",
      ],
      "type": "string",
    },
    "bool": {
      "type": "boolean",
    },
    "dict": {
      "additionalProperties": {
        "anyOf": [
          {
            "type": "string",
          },
        ],
      },
      "type": "object",
    },
    "enum": {
      "anyOf": [
        {
          "enum": [
            "A",
          ],
          "title": "A",
          "type": "string",
        },
        {
          "enum": [
            "B",
          ],
          "title": "B",
          "type": "string",
        },
        {
          "enum": [
            "C",
          ],
          "title": "C",
          "type": "string",
        },
      ],
    },
    "literalBoolean": {
      "enum": [
        true,
      ],
      "type": "boolean",
    },
    "literalNumber": {
      "enum": [
        1,
      ],
      "type": "number",
    },
    "literalString": {
      "enum": [
        "literal",
      ],
      "type": "string",
    },
    "number": {
      "type": "number",
    },
    "oneof": {
      "anyOf": [
        {
          "properties": {
            "foo": {
              "type": "string",
            },
          },
          "required": [
            "foo",
          ],
          "type": "object",
        },
        {
          "properties": {
            "bar": {
              "type": "string",
            },
          },
          "required": [
            "bar",
          ],
          "type": "object",
        },
      ],
    },
    "optionalSelfCopy": {
      "properties": {
        "bMember": {
          "enum": [
            "B",
          ],
          "type": "string",
        },
        "bool": {
          "type": "boolean",
        },
        "dict": {
          "additionalProperties": {
            "anyOf": [
              {
                "type": "string",
              },
            ],
          },
          "type": "object",
        },
        "enum": {
          "anyOf": [
            {
              "enum": [
                "A",
              ],
              "title": "A",
              "type": "string",
            },
            {
              "enum": [
                "B",
              ],
              "title": "B",
              "type": "string",
            },
            {
              "enum": [
                "C",
              ],
              "title": "C",
              "type": "string",
            },
          ],
        },
        "id": {
          "type": "string",
        },
        "literalBoolean": {
          "enum": [
            true,
          ],
          "type": "boolean",
        },
        "literalNumber": {
          "enum": [
            1,
          ],
          "type": "number",
        },
        "literalString": {
          "enum": [
            "literal",
          ],
          "type": "string",
        },
        "number": {
          "type": "number",
        },
        "oneof": {
          "anyOf": [
            {
              "properties": {
                "foo": {
                  "type": "string",
                },
              },
              "required": [
                "foo",
              ],
              "type": "object",
            },
            {
              "properties": {
                "bar": {
                  "type": "string",
                },
              },
              "required": [
                "bar",
              ],
              "type": "object",
            },
          ],
        },
        "recordIntersection": {
          "allOf": [
            {
              "properties": {
                "a": {
                  "type": "string",
                },
              },
              "required": [
                "a",
              ],
              "type": "object",
            },
            {
              "properties": {
                "b": {
                  "type": "number",
                },
              },
              "required": [
                "b",
              ],
              "type": "object",
            },
          ],
        },
        "stringArray": {
          "items": [
            {
              "type": "string",
            },
          ],
          "type": "array",
        },
        "tuple": {
          "items": [
            {
              "properties": {
                "id": {
                  "type": "string",
                },
              },
              "required": [],
              "type": "object",
            },
            {
              "enum": [
                "separator",
              ],
              "type": "string",
            },
            {
              "type": "boolean",
            },
          ],
          "required": [
            "0",
            "1",
            "2",
          ],
          "type": "array",
        },
        "unionArray": {
          "items": [
            {
              "properties": {
                "foo": {
                  "type": "string",
                },
              },
              "required": [
                "foo",
              ],
              "type": "object",
            },
            {
              "properties": {
                "bar": {
                  "type": "string",
                },
              },
              "required": [
                "bar",
              ],
              "type": "object",
            },
          ],
          "type": "array",
        },
      },
      "required": [],
      "type": "object",
    },
    "recordIntersection": {
      "allOf": [
        {
          "properties": {
            "a": {
              "type": "string",
            },
          },
          "required": [
            "a",
          ],
          "type": "object",
        },
        {
          "properties": {
            "b": {
              "type": "number",
            },
          },
          "required": [
            "b",
          ],
          "type": "object",
        },
      ],
    },
    "stringArray": {
      "items": [
        {
          "type": "string",
        },
      ],
      "type": "array",
    },
    "tuple": {
      "items": [
        {
          "type": "string",
        },
        {
          "type": "number",
        },
      ],
      "required": [
        "0",
        "1",
      ],
      "type": "array",
    },
    "unionArray": {
      "items": [
        {
          "properties": {
            "foo": {
              "type": "string",
            },
          },
          "required": [
            "foo",
          ],
          "type": "object",
        },
        {
          "properties": {
            "bar": {
              "type": "string",
            },
          },
          "required": [
            "bar",
          ],
          "type": "object",
        },
      ],
      "type": "array",
    },
    "uuid": {
      "type": "string",
    },
  },
  "required": [
    "uuid",
    "number",
    "bool",
    "stringArray",
    "unionArray",
    "tuple",
    "literalString",
    "literalNumber",
    "literalBoolean",
    "enum",
    "bMember",
    "recordIntersection",
    "dict",
    "oneof",
  ],
  "type": "object",
}
`;

exports[`toJsonSchema should parse to a valid JSONSchema when set to 'set-as-any' 1`] = `
{
  "properties": {
    "bMember": {
      "enum": [
        "B",
      ],
      "type": "string",
    },
    "bool": {
      "type": "boolean",
    },
    "customValidator": {
      "title": "Custom Validator (anonymous)",
    },
    "dict": {
      "additionalProperties": {
        "anyOf": [
          {
            "type": "string",
          },
        ],
      },
      "type": "object",
    },
    "enum": {
      "anyOf": [
        {
          "enum": [
            "A",
          ],
          "title": "A",
          "type": "string",
        },
        {
          "enum": [
            "B",
          ],
          "title": "B",
          "type": "string",
        },
        {
          "enum": [
            "C",
          ],
          "title": "C",
          "type": "string",
        },
      ],
    },
    "literalBoolean": {
      "enum": [
        true,
      ],
      "type": "boolean",
    },
    "literalNumber": {
      "enum": [
        1,
      ],
      "type": "number",
    },
    "literalString": {
      "enum": [
        "literal",
      ],
      "type": "string",
    },
    "number": {
      "type": "number",
    },
    "oneof": {
      "anyOf": [
        {
          "properties": {
            "foo": {
              "type": "string",
            },
          },
          "required": [
            "foo",
          ],
          "type": "object",
        },
        {
          "properties": {
            "bar": {
              "type": "string",
            },
          },
          "required": [
            "bar",
          ],
          "type": "object",
        },
      ],
    },
    "optionalSelfCopy": {
      "properties": {
        "bMember": {
          "enum": [
            "B",
          ],
          "type": "string",
        },
        "bool": {
          "type": "boolean",
        },
        "dict": {
          "additionalProperties": {
            "anyOf": [
              {
                "type": "string",
              },
            ],
          },
          "type": "object",
        },
        "enum": {
          "anyOf": [
            {
              "enum": [
                "A",
              ],
              "title": "A",
              "type": "string",
            },
            {
              "enum": [
                "B",
              ],
              "title": "B",
              "type": "string",
            },
            {
              "enum": [
                "C",
              ],
              "title": "C",
              "type": "string",
            },
          ],
        },
        "id": {
          "type": "string",
        },
        "literalBoolean": {
          "enum": [
            true,
          ],
          "type": "boolean",
        },
        "literalNumber": {
          "enum": [
            1,
          ],
          "type": "number",
        },
        "literalString": {
          "enum": [
            "literal",
          ],
          "type": "string",
        },
        "number": {
          "type": "number",
        },
        "oneof": {
          "anyOf": [
            {
              "properties": {
                "foo": {
                  "type": "string",
                },
              },
              "required": [
                "foo",
              ],
              "type": "object",
            },
            {
              "properties": {
                "bar": {
                  "type": "string",
                },
              },
              "required": [
                "bar",
              ],
              "type": "object",
            },
          ],
        },
        "recordIntersection": {
          "allOf": [
            {
              "properties": {
                "a": {
                  "type": "string",
                },
              },
              "required": [
                "a",
              ],
              "type": "object",
            },
            {
              "properties": {
                "b": {
                  "type": "number",
                },
              },
              "required": [
                "b",
              ],
              "type": "object",
            },
          ],
        },
        "stringArray": {
          "items": [
            {
              "type": "string",
            },
          ],
          "type": "array",
        },
        "stringSet": {
          "title": "Set",
        },
        "tuple": {
          "items": [
            {
              "properties": {
                "id": {
                  "type": "string",
                },
              },
              "required": [],
              "type": "object",
            },
            {
              "enum": [
                "separator",
              ],
              "type": "string",
            },
            {
              "type": "boolean",
            },
          ],
          "required": [
            "0",
            "1",
            "2",
          ],
          "type": "array",
        },
        "unionArray": {
          "items": [
            {
              "properties": {
                "foo": {
                  "type": "string",
                },
              },
              "required": [
                "foo",
              ],
              "type": "object",
            },
            {
              "properties": {
                "bar": {
                  "type": "string",
                },
              },
              "required": [
                "bar",
              ],
              "type": "object",
            },
          ],
          "type": "array",
        },
      },
      "required": [],
      "type": "object",
    },
    "recordIntersection": {
      "allOf": [
        {
          "properties": {
            "a": {
              "type": "string",
            },
          },
          "required": [
            "a",
          ],
          "type": "object",
        },
        {
          "properties": {
            "b": {
              "type": "number",
            },
          },
          "required": [
            "b",
          ],
          "type": "object",
        },
      ],
    },
    "stringArray": {
      "items": [
        {
          "type": "string",
        },
      ],
      "type": "array",
    },
    "stringSet": {
      "title": "Set",
    },
    "symbol": {
      "title": "Symbol",
    },
    "tuple": {
      "items": [
        {
          "type": "string",
        },
        {
          "type": "number",
        },
      ],
      "required": [
        "0",
        "1",
      ],
      "type": "array",
    },
    "undef": {
      "title": "undefined",
    },
    "unionArray": {
      "items": [
        {
          "properties": {
            "foo": {
              "type": "string",
            },
          },
          "required": [
            "foo",
          ],
          "type": "object",
        },
        {
          "properties": {
            "bar": {
              "type": "string",
            },
          },
          "required": [
            "bar",
          ],
          "type": "object",
        },
      ],
      "type": "array",
    },
    "uuid": {
      "type": "string",
    },
  },
  "required": [
    "uuid",
    "number",
    "bool",
    "stringSet",
    "stringArray",
    "unionArray",
    "tuple",
    "literalString",
    "literalNumber",
    "literalBoolean",
    "enum",
    "bMember",
    "recordIntersection",
    "dict",
    "oneof",
    "symbol",
    "undef",
    "customValidator",
  ],
  "type": "object",
}
`;

exports[`toJsonSchema should set the specified additionalProperty 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "bMember": {
      "enum": [
        "B",
      ],
      "type": "string",
    },
    "bool": {
      "type": "boolean",
    },
    "dict": {
      "additionalProperties": {
        "anyOf": [
          {
            "type": "string",
          },
        ],
      },
      "type": "object",
    },
    "enum": {
      "anyOf": [
        {
          "enum": [
            "A",
          ],
          "title": "A",
          "type": "string",
        },
        {
          "enum": [
            "B",
          ],
          "title": "B",
          "type": "string",
        },
        {
          "enum": [
            "C",
          ],
          "title": "C",
          "type": "string",
        },
      ],
    },
    "literalBoolean": {
      "enum": [
        true,
      ],
      "type": "boolean",
    },
    "literalNumber": {
      "enum": [
        1,
      ],
      "type": "number",
    },
    "literalString": {
      "enum": [
        "literal",
      ],
      "type": "string",
    },
    "number": {
      "type": "number",
    },
    "oneof": {
      "anyOf": [
        {
          "additionalProperties": false,
          "properties": {
            "foo": {
              "type": "string",
            },
          },
          "required": [
            "foo",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "properties": {
            "bar": {
              "type": "string",
            },
          },
          "required": [
            "bar",
          ],
          "type": "object",
        },
      ],
    },
    "optionalSelfCopy": {
      "additionalProperties": false,
      "properties": {
        "bMember": {
          "enum": [
            "B",
          ],
          "type": "string",
        },
        "bool": {
          "type": "boolean",
        },
        "dict": {
          "additionalProperties": {
            "anyOf": [
              {
                "type": "string",
              },
            ],
          },
          "type": "object",
        },
        "enum": {
          "anyOf": [
            {
              "enum": [
                "A",
              ],
              "title": "A",
              "type": "string",
            },
            {
              "enum": [
                "B",
              ],
              "title": "B",
              "type": "string",
            },
            {
              "enum": [
                "C",
              ],
              "title": "C",
              "type": "string",
            },
          ],
        },
        "id": {
          "type": "string",
        },
        "literalBoolean": {
          "enum": [
            true,
          ],
          "type": "boolean",
        },
        "literalNumber": {
          "enum": [
            1,
          ],
          "type": "number",
        },
        "literalString": {
          "enum": [
            "literal",
          ],
          "type": "string",
        },
        "number": {
          "type": "number",
        },
        "oneof": {
          "anyOf": [
            {
              "additionalProperties": false,
              "properties": {
                "foo": {
                  "type": "string",
                },
              },
              "required": [
                "foo",
              ],
              "type": "object",
            },
            {
              "additionalProperties": false,
              "properties": {
                "bar": {
                  "type": "string",
                },
              },
              "required": [
                "bar",
              ],
              "type": "object",
            },
          ],
        },
        "recordIntersection": {
          "allOf": [
            {
              "additionalProperties": false,
              "properties": {
                "a": {
                  "type": "string",
                },
              },
              "required": [
                "a",
              ],
              "type": "object",
            },
            {
              "additionalProperties": false,
              "properties": {
                "b": {
                  "type": "number",
                },
              },
              "required": [
                "b",
              ],
              "type": "object",
            },
          ],
        },
        "stringArray": {
          "items": [
            {
              "type": "string",
            },
          ],
          "type": "array",
        },
        "tuple": {
          "items": [
            {
              "additionalProperties": false,
              "properties": {
                "id": {
                  "type": "string",
                },
              },
              "required": [],
              "type": "object",
            },
            {
              "enum": [
                "separator",
              ],
              "type": "string",
            },
            {
              "type": "boolean",
            },
          ],
          "required": [
            "0",
            "1",
            "2",
          ],
          "type": "array",
        },
        "unionArray": {
          "items": [
            {
              "additionalProperties": false,
              "properties": {
                "foo": {
                  "type": "string",
                },
              },
              "required": [
                "foo",
              ],
              "type": "object",
            },
            {
              "additionalProperties": false,
              "properties": {
                "bar": {
                  "type": "string",
                },
              },
              "required": [
                "bar",
              ],
              "type": "object",
            },
          ],
          "type": "array",
        },
      },
      "required": [],
      "type": "object",
    },
    "recordIntersection": {
      "allOf": [
        {
          "additionalProperties": false,
          "properties": {
            "a": {
              "type": "string",
            },
          },
          "required": [
            "a",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "properties": {
            "b": {
              "type": "number",
            },
          },
          "required": [
            "b",
          ],
          "type": "object",
        },
      ],
    },
    "stringArray": {
      "items": [
        {
          "type": "string",
        },
      ],
      "type": "array",
    },
    "tuple": {
      "items": [
        {
          "type": "string",
        },
        {
          "type": "number",
        },
      ],
      "required": [
        "0",
        "1",
      ],
      "type": "array",
    },
    "unionArray": {
      "items": [
        {
          "additionalProperties": false,
          "properties": {
            "foo": {
              "type": "string",
            },
          },
          "required": [
            "foo",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "properties": {
            "bar": {
              "type": "string",
            },
          },
          "required": [
            "bar",
          ],
          "type": "object",
        },
      ],
      "type": "array",
    },
    "uuid": {
      "type": "string",
    },
  },
  "required": [
    "uuid",
    "number",
    "bool",
    "stringArray",
    "unionArray",
    "tuple",
    "literalString",
    "literalNumber",
    "literalBoolean",
    "enum",
    "bMember",
    "recordIntersection",
    "dict",
    "oneof",
  ],
  "type": "object",
}
`;

exports[`toJsonSchema should use the custom parsers 1`] = `
{
  "properties": {
    "custom": {
      "title": "Custom parser schema for validateFunction",
    },
    "function": {
      "title": "Custom parser schema for Function DataType",
    },
    "set": {
      "title": "Custom parser schema for Set DataType",
    },
    "symbol": {
      "title": "Custom parser schema for Symbol DataType",
    },
    "undefined": {
      "title": "Custom parser schema for Undefined DataType",
    },
  },
  "required": [
    "custom",
    "function",
    "set",
    "symbol",
    "undefined",
  ],
  "type": "object",
}
`;
